src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ClusterApplierServiceStatsCollector.java

```diff
@@ -0,0 +1,190 @@
+        value = new StringBuilder();
+        this.controller = controller;
+        this.configOverridesWrapper = configOverridesWrapper;
+        this.metricCache = new MetricCache();
+    }
+
+    @Override
+    public void collectMetrics(long startTime) {
+        try {
+            long mCurrT = System.currentTimeMillis();
+            if (ESResources.INSTANCE.getClusterService() == null
+                    || ESResources.INSTANCE.getClusterService().getClusterApplierService() == null) {
+                return;
+            }
+            mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+            mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+            ClusterApplierServiceStats metricsObject = null;
+            try {
+                metricsObject = mapper.readValue(
+                        mapper.writeValueAsString(getClusterApplierServiceStats()), ClusterApplierServiceStats.class);
+            } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException ex) {
+                LOG.debug("No method found to get cluster state applier thread stats. " +
+                        "Skipping ClusterApplierServiceStatsCollector");
+                return;
+            }
+
+            Long latencyInMillis = computeLatency(metricsObject);
+            Long failure = computeFailure(metricsObject);
+
+            if(latencyInMillis != null && failure != null) {
+                ClusterApplierServiceMetrics clusterApplierServiceMetrics = new
+                        ClusterApplierServiceMetrics(latencyInMillis, failure);
+
+                value.setLength(0);
+                value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds())
+                        .append(PerformanceAnalyzerMetrics.sMetricNewLineDelimitor);
+                value.append(clusterApplierServiceMetrics.serialize());
+                saveMetricValues(value.toString(), startTime);
+            }
+            this.metricCache = new MetricCache(metricsObject.timeTakenInMillis, metricsObject.failedCount,
+                    metricsObject.totalCount);
+            PerformanceAnalyzerApp.WRITER_METRICS_AGGREGATOR.updateStat(
+                    WriterMetrics.CLUSTER_APPLIER_SERVICE_STATS_COLLECTOR_EXECUTION_TIME, "",
+                    System.currentTimeMillis() - mCurrT);
+        } catch (Exception ex) {
+            PerformanceAnalyzerApp.ERRORS_AND_EXCEPTIONS_AGGREGATOR.updateStat(
+                    ExceptionsAndErrors.CLUSTER_APPLIER_SERVICE_STATS_COLLECTOR_ERROR, "", 1);
+            LOG.debug("Exception in Collecting Cluster Applier Service Metrics: {} for startTime {}",
+                    () -> ex.toString(), () -> startTime);
+        }
+    }
+
+    private Long computeLatency(final ClusterApplierServiceStats metricsObject) {
+        if(metricCache.prevTimeTakenInMillis == null || metricCache.prevTotalCount == null) {
```

Comment:
Since you are initializing the metricCache object with null values in class constructor, the first call to this method should give wrong result (return null when it should not). Am I missing something here?

Reply:
Changed implementation to persist the first value as well.